File: test_bmi_calculator.py
import unittest

from BMI_Calculator import bmi_category, calculate_bmi


class BmiCalculatorTests(unittest.TestCase):
    def test_calculation(self) -> None:
        self.assertAlmostEqual(calculate_bmi(70, 1.75), 22.8571428571)

    def test_categories_at_boundaries(self) -> None:
        cases = {
            18.49: "Underweight",
            18.5: "Normal weight",
            24.99: "Normal weight",
            25.0: "Overweight",
            29.99: "Overweight",
            30.0: "Obese",
        }
        for bmi, expected in cases.items():
            with self.subTest(bmi=bmi):
                self.assertEqual(bmi_category(bmi), expected)

    def test_invalid_measurements(self) -> None:
        for mass, height in ((0, 1.7), (70, 0), (-1, 1.7), (70, -1.7)):
            with self.subTest(mass=mass, height=height):
                with self.assertRaises(ValueError):
                    calculate_bmi(mass, height)


if __name__ == "__main__":
    unittest.main()

File: test_bmi_calculator.c
#include <assert.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

bool calculate_bmi(double mass_kg, double height_m, double *result);
const char *bmi_category(double bmi);

static bool close_enough(double left, double right)
{
    double difference = left - right;
    if (difference < 0.0) {
        difference = -difference;
    }
    return difference < 0.000001;
}

int main(void)
{
    double result = 0.0;

    assert(calculate_bmi(70.0, 1.75, &result));
    assert(close_enough(result, 22.8571428571));
    assert(!calculate_bmi(70.0, 0.0, &result));
    assert(!calculate_bmi(-1.0, 1.75, &result));
    assert(strcmp(bmi_category(18.49), "Underweight") == 0);
    assert(strcmp(bmi_category(18.5), "Normal weight") == 0);
    assert(strcmp(bmi_category(25.0), "Overweight") == 0);
    assert(strcmp(bmi_category(30.0), "Obese") == 0);

    puts("C tests passed.");
    return 0;
}
